code/game/objects/items/robot/robot_upgrades.dm

```diff
@@ -721,3 +721,103 @@
 		var/obj/item/borg/apparatus/beaker/extra/E = locate() in R.module.modules
 		if (E)
 			R.module.remove_module(E, TRUE)
+
+
+/obj/item/borg/upgrade/speciality
+	name = "Speciality Module"
+	icon_state = "cyborg_upgrade3"
+	require_module = TRUE
+	module_type = /obj/item/robot_module/butler
+	var/obj/item/hat
+	var/addmodules = list()
+	var/list/additional_reagents = list()
+
+/obj/item/borg/upgrade/speciality/action(mob/living/silicon/robot/R, user = usr)
+	. = ..()
+	if(.)
+		for(var/obj/item/borg/upgrade/SPEC in R.upgrades)
+			if (istype(SPEC,/obj/item/borg/upgrade/speciality) && SPEC != src)
+				SPEC.deactivate(R)
+				R.upgrades -= SPEC
+				qdel(SPEC)
+
+
+		for(var/module in src.addmodules)
+			var/obj/item/nmodule = locate(module) in R
+			if (!nmodule)
+				nmodule = new module(R.module)
+				R.module.basic_modules += nmodule
+				R.module.add_module(nmodule, FALSE, TRUE)
+				
+		for(var/obj/item/reagent_containers/borghypo/borgshaker/H in R.module.modules)
+			for(var/re in additional_reagents)
+				H.add_reagent(re)
+
+		if(hat && R.hat_offset != INFINITY && !R.hat)
+			var/obj/item/equipt = new hat(src)
+			if (equipt )
+				R.place_on_head(equipt)
+	return .
+
+/obj/item/borg/upgrade/speciality/deactivate(mob/living/silicon/robot/R, user = usr)
+	. = ..()
+	if (.)
+		//Remove existing modules indiscriminately
+		for(var/module in src.addmodules)
+			var/dmod = locate(module) in R.module.modules
+			if (dmod)
+				R.module.remove_module(dmod, TRUE)
+		for(var/obj/item/reagent_containers/borghypo/borgshaker/H in R.module.modules)
+			for(var/re in additional_reagents)
+				H.del_reagent(re)
+	return .
+
+/obj/item/borg/upgrade/speciality/kitchen
```

Comment:
The lists are inconsistent, sometimes the last element of the list has a comma after it,  but sometimes it doesn't. Either way is fine but do it the same for all of them